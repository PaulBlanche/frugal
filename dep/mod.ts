import './std/fmt/bytes.ts';
import './std/fmt/colors.ts';
import './std/http/file_server.ts';
import './std/http/http_status.ts';
import './std/testing/asserts.ts';
import './std/testing/mock.ts';
import './std/testing/snapshot.ts';
import './std/testing/time.ts';
import './std/dotenv.ts';
import './std/flags.ts';
import './std/fs.ts';
import './std/http.ts';
import './std/path.ts';
import './std/streams.ts';
import './cheerio.ts';
import './deno_graph.ts';
import './deno-dom-wasm.ts';
import './djwt.ts';
import './esbuild.ts';
import './importmap.ts';
import './lightningcss.ts';
import './path-to-regexp.ts';
import './puppeteer/mod.ts';
import './puppeteer/install.ts';
import './xxhash.ts';
import './zod.ts';

import 'preact';
import 'preact/hooks';
import 'preact/jsx-runtime';
import 'preact-render-to-string';
