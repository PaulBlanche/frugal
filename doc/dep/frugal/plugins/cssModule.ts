export * from "frugal/plugins/cssModule.ts";
