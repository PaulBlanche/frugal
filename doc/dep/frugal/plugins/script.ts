export * from 'frugal/plugins/script.ts';
