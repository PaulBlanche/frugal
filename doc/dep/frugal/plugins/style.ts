export * from 'frugal/plugins/style.ts';
