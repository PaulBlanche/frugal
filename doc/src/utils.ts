type ThrottleOptions = {
  timeout: number;
  leading?: boolean;
  trailing?: boolean;
};

export function throttle<ARGS extends any[]>(
  callback: (...args: ARGS) => void,
  { timeout, leading = false, trailing = false }: ThrottleOptions,
): (...args: ARGS) => void {
  let timeoutHandle: number | undefined = undefined;
  let previous = 0;

  return (...args) => {
    const now = Date.now();

    if (!previous && !leading) {
      previous = now;
    }

    const remaining = timeout - (now - previous);

    if (remaining <= 0) {
      clearTimeout(timeoutHandle);
      timeoutHandle = undefined;
      previous = now;
      callback(...args);
    } else if (timeoutHandle === undefined && trailing) {
      timeoutHandle = setTimeout(() => {
        previous = Date.now();
        timeoutHandle = undefined;
        callback(...args);
      }, remaining);
    }
  };
}
