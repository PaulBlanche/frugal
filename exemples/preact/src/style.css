html,
body {
    font-family: system-ui;
    margin: 0;
}

body {
    padding: 2rem;
}

main {
    max-width: 600px;
    margin: 0 auto;
}

.scroll-padder {
    padding: 200px 0 400px;
    text-align: center;
}