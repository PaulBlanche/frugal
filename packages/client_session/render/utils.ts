import { NodeType } from './types.ts';

export function clone(node: Node) {
    if (isScriptElement(node)) {
        // for scripts cloning is not enough, we need to recreate the script
        // element for it to be loaded/evaluated again
        const script = document.createElement('script');
        for (const attribute of node.attributes) {
            script.setAttribute(attribute.name, attribute.value);
        }
        script.innerHTML = node.innerHTML;
        return script;
    }
    return node.cloneNode(true);
}

function isScriptElement(node: Node): node is HTMLScriptElement {
    return isElement(node) && node.nodeName === 'SCRIPT';
}

export function isElement(node: Node): node is Element {
    return node.nodeType === NodeType.ELEMENT_NODE;
}

export function hash(
    element: Element,
    attributeFilter?: string[],
): string {
    const key = [];

    if (attributeFilter) {
        for (const name of attributeFilter) {
            const value = element.getAttribute(name);
            if (value) {
                key.push(`${name}="${element.getAttribute(name)}"`);
            }
        }
    } else {
        for (const attribute of element.attributes) {
            key.push(`${attribute.name}="${attribute.value}"`);
        }
    }

    key.sort();

    return `${element.tagName} ${key.join(' ')} ${element.innerHTML}`;
}

type Attribute = [name: string, value: string | boolean];

const BOOLEAN_ATTRIBUTES = [
    'allowfullscreen',
    'async',
    'autofocus',
    'autoplay',
    'checked',
    'controls',
    'default',
    'defer',
    'disabled',
    'formnovalidate',
    'inert',
    'ismap',
    'itemscope',
    'loop',
    'multiple',
    'muted',
    'nomodule',
    'novalidate',
    'open',
    'playsinline',
    'readonly',
    'required',
    'reversed',
    'selected',
];

export function getAttributes(element: Element): Attribute[] {
    const attributes: Record<string, string | boolean> = {};

    for (const { name, value } of element.attributes) {
        attributes[name] = value;
    }

    for (const name of BOOLEAN_ATTRIBUTES) {
        attributes[name] = element.hasAttribute(name);
    }

    return Object.entries(attributes);
}
