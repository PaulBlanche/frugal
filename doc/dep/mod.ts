import './frugal/plugins/cssModule.ts';
import './frugal/plugins/script.ts';
import './frugal/plugins/style.ts';
import './frugal/plugins/svg.ts';
import './frugal/runtime/client_session.ts';
import './frugal/runtime/preact.client.ts';
import './frugal/runtime/preact.server.ts';
import './frugal/mod.ts';
import './frugal/page.ts';
import './frugal/persistence.ts';
import './std/dotenv.ts';
import './clsx.ts';
import './he.ts';
import './prism.ts';

import 'preact';
import 'preact/hooks';
import 'preact/jsx-runtime';
import 'preact-render-to-string';
