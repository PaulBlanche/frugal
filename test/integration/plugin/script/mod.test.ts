import * as asserts from "../../../../dep/std/testing/asserts.ts";
import * as snapshot from "../../../../dep/std/testing/snapshot.ts";
import { Config, context } from "../../../../mod.ts";
import { BuildHelper, setupFiles } from "../../../utils.ts";

if (import.meta.main) {
    const config = await loadConfig();
    await context(config).watch();
} else {
    await setupTestFiles();
}

Deno.test("script: build page with script and css module dependencies", async (t) => {
    const config = await loadConfig();
    const helper = new BuildHelper(config);

    await helper.build();

    snapshot.assertSnapshot(t, await Deno.readTextFile(new URL("css/page.css", helper.config.publicdir)));
    snapshot.assertSnapshot(t, await Deno.readTextFile(new URL("js/page.js", helper.config.publicdir)));
});

Deno.test("css: script dependencies are watched", async () => {
    const config = await loadConfig();
    const helper = new BuildHelper(config);

    const context = helper.context();
    context.watch();

    await context.awaitNextBuild();

    const firstBuildCache = await context.cacheExplorer();
    const updatedAt = firstBuildCache.get("/page").updatedAt;

    // add a comment at the top of dep.css
    const depModuleURL = new URL("./project/dep.ts", import.meta.url);
    const originalData = await Deno.readTextFile(depModuleURL);
    await Deno.writeTextFile(depModuleURL, `//comment\n${originalData}`);

    await context.awaitNextBuild();

    const secondBuildChache = await context.cacheExplorer();
    asserts.assertNotEquals(updatedAt, secondBuildChache.get("/page").updatedAt);

    await Deno.writeTextFile(depModuleURL, originalData);

    await context.dispose();
});

async function setupTestFiles() {
    // clean everything from previous tests
    const base = new URL("./project/", import.meta.url);
    try {
        await Deno.remove(base, { recursive: true });
    } catch {}

    // setup clean files for current tests
    await setupFiles(base, {
        //####
        "./style.module.css": `.foo { color: red; }
`,

        //####
        "./dep1.script.ts": `import "./shared.ts";

if (import.meta.main) {
    console.log("dep1.script.ts")
}
`,

        //####
        "./dep2.script.ts": `import "./shared.ts";

if (import.meta.main) {
    console.log("dep2.script.ts")
}
`,

        //####
        "./dep.ts": `import "./dep1.script.ts";
import "./dep2.script.ts";
`,

        //####
        "./shared.ts": `import 'npm:pad-left@2.1.0';
import 'https://esm.sh/fast-cartesian@8.0.0';
import style from './style.module.css'

// side effect to keep style import
window.style = style
`,

        //####
        "./before.script.ts": `import "./shared.ts";

if (import.meta.main) {
    console.log("before.script.ts")
}
`,

        //####
        "./after.script.ts": `import "./shared.ts";

if (import.meta.main) {
    console.log("after.script.ts")
}
`,

        //####
        "./page.ts": `import "./before.script.ts";
import "./dep.ts";
import "./after.script.ts";

export const pattern = "/page";

export function render() {
    return "";
}
`,

        //####
        "./frugal.config.ts": `import { Config } from "../../../../../mod.ts";
import { script } from "../../../../../plugins/script.ts";
import { cssModule } from "../../../../../plugins/cssModule.ts";
import { css } from "../../../../../plugins/css.ts";

export const config: Config = {
    self: import.meta.url,
    outdir: "./dist/",
    pages: ["./page.ts"],
    plugins: [
        script(), 
        cssModule({
            // to avoid hash beeing different on different machines runing the
            // tests
            pattern: "[local]",
        }), 
        css(),
    ],
    log: { level: "silent" },
};
`,
    });
}

async function loadConfig(): Promise<Config> {
    // load config busting deno cache
    const hash = String(Date.now());
    const { config } = await import(`./project/frugal.config.ts#${hash}`);
    return config;
}
