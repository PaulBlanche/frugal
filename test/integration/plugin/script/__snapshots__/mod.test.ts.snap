export const snapshot = {};

snapshot[`script: build page with script and css module dependencies 1`] = `
"/* virtual:/style.module.css-68ec91d547ef2ffa.frugal-compiled-module.css */
" +
  ".foo {
" +
  "  color: red;
" +
  "}
"
`;

snapshot[`script: build page with script and css module dependencies 2`] = `
"var __create = Object.create;
" +
  "var __defProp = Object.defineProperty;
" +
  "var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
" +
  "var __getOwnPropNames = Object.getOwnPropertyNames;
" +
  "var __getProtoOf = Object.getPrototypeOf;
" +
  "var __hasOwnProp = Object.prototype.hasOwnProperty;
" +
  "var __commonJS = (cb, mod) => function __require() {
" +
  "  return mod || (0, cb[__getOwnPropNames(cb)[0]])((mod = { exports: {} }).exports, mod), mod.exports;
" +
  "};
" +
  "var __copyProps = (to, from, except, desc) => {
" +
  '  if (from && typeof from === "object" || typeof from === "function") {
' +
  "    for (let key of __getOwnPropNames(from))
" +
  "      if (!__hasOwnProp.call(to, key) && key !== except)
" +
  "        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
" +
  "  }
" +
  "  return to;
" +
  "};
" +
  "var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
" +
  "  // If the importer is in node compatibility mode or this is not an ESM
" +
  "  // file that has been converted to a CommonJS file using a Babel-
" +
  '  // compatible transform (i.e. "__esModule" has not been set), then set
' +
  '  // "default" to the CommonJS "module.exports" for node compatibility.
' +
  '  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
' +
  "  mod
" +
  "));
" +
  "
" +
  "// node_modules/.deno/repeat-string@1.6.1/node_modules/repeat-string/index.js
" +
  "var require_repeat_string = __commonJS({
" +
  '  "node_modules/.deno/repeat-string@1.6.1/node_modules/repeat-string/index.js"(exports, module) {
' +
  '    "use strict";
' +
  '    var res = "";
' +
  "    var cache;
" +
  "    module.exports = repeat;
" +
  "    function repeat(str, num) {
" +
  '      if (typeof str !== "string") {
' +
  '        throw new TypeError("expected a string");
' +
  "      }
" +
  "      if (num === 1)
" +
  "        return str;
" +
  "      if (num === 2)
" +
  "        return str + str;
" +
  "      var max = str.length * num;
" +
  '      if (cache !== str || typeof cache === "undefined") {
' +
  "        cache = str;
" +
  '        res = "";
' +
  "      } else if (res.length >= max) {
" +
  "        return res.substr(0, max);
" +
  "      }
" +
  "      while (max > res.length && num > 1) {
" +
  "        if (num & 1) {
" +
  "          res += str;
" +
  "        }
" +
  "        num >>= 1;
" +
  "        str += str;
" +
  "      }
" +
  "      res += str;
" +
  "      res = res.substr(0, max);
" +
  "      return res;
" +
  "    }
" +
  "  }
" +
  "});
" +
  "
" +
  "// node_modules/.deno/pad-left@2.1.0/node_modules/pad-left/index.js
" +
  "var require_pad_left = __commonJS({
" +
  '  "node_modules/.deno/pad-left@2.1.0/node_modules/pad-left/index.js"(exports, module) {
' +
  '    "use strict";
' +
  "    var repeat = require_repeat_string();
" +
  "    module.exports = function padLeft(str, num, ch) {
" +
  "      str = str.toString();
" +
  '      if (typeof num === "undefined") {
' +
  "        return str;
" +
  "      }
" +
  "      if (ch === 0) {
" +
  '        ch = "0";
' +
  "      } else if (ch) {
" +
  "        ch = ch.toString();
" +
  "      } else {
" +
  '        ch = " ";
' +
  "      }
" +
  "      return repeat(ch, num - str.length) + str;
" +
  "    };
" +
  "  }
" +
  "});
" +
  "
" +
  "// shared.ts
" +
  "var import_npm_pad_left_2_1 = __toESM(require_pad_left());
" +
  "
" +
  "// style.module.css
" +
  "var style_module_default = {
" +
  '  "foo": Array.from(/* @__PURE__ */ new Set(["foo"]))
' +
  "};
" +
  "
" +
  "// shared.ts
" +
  "window.style = style_module_default;
" +
  "
" +
  "// before.script.ts
" +
  "if (true) {
" +
  '  console.log("before.script.ts");
' +
  "}
" +
  "
" +
  "// dep1.script.ts
" +
  "if (true) {
" +
  '  console.log("dep1.script.ts");
' +
  "}
" +
  "
" +
  "// dep2.script.ts
" +
  "if (true) {
" +
  '  console.log("dep2.script.ts");
' +
  "}
" +
  "
" +
  "// after.script.ts
" +
  "if (true) {
" +
  '  console.log("after.script.ts");
' +
  "}
" +
  "/*! Bundled license information:
" +
  "
" +
  "repeat-string/index.js:
" +
  "  (*!
" +
  "   * repeat-string <https://github.com/jonschlinkert/repeat-string>
" +
  "   *
" +
  "   * Copyright (c) 2014-2015, Jon Schlinkert.
" +
  "   * Licensed under the MIT License.
" +
  "   *)
" +
  "
" +
  "pad-left/index.js:
" +
  "  (*!
" +
  "   * pad-left <https://github.com/jonschlinkert/pad-left>
" +
  "   *
" +
  "   * Copyright (c) 2014-2015, Jon Schlinkert.
" +
  "   * Licensed under the MIT license.
" +
  "   *)
" +
  "*/
"
`;
